datastream-bigquery/src/main/java/com/linkedin/datastream/bigquery/BigqueryBatchCommitter.java

```diff
@@ -195,80 +205,78 @@ public void commit(List<InsertAllRequest.RowToInsert> batch,
+
+            final InsertAllRequest insertAllRequest = InsertAllRequest.newBuilder(tableId, batch).build();
+
+            final long start = System.currentTimeMillis();
+            Map<Integer, Exception> insertErrors = insertRowsAndMapErrors(insertAllRequest);
+            final long end = System.currentTimeMillis();
+            DynamicMetricsManager.getInstance().createOrUpdateHistogram(classSimpleName, recordMetadata.get(0).getTopic(),
+                    "insertAllExecTime", end - start);
+
+            // If we encountered insert errors, try creating/updating the destination table before retrying
+            if (!insertErrors.isEmpty()) {
+                try {
+                    final boolean tableUpdatedOrCreated = createOrUpdateTable(destination);
+                    if (tableUpdatedOrCreated) {
+                        LOG.info("Table created/updated for destination {}. Retrying batch...", destination);
+                        insertErrors = insertRowsAndMapErrors(insertAllRequest);
                     }
+                } catch (final Exception e) {
+                    final DatastreamTransientException wrappedException = new DatastreamTransientException(e);
+                    insertErrors = IntStream.range(0, batch.size()).boxed().collect(Collectors.toMap(i -> i, i -> wrappedException));
                 }
             }
 
+            final Map<Integer, Exception> finalInsertErrors = insertErrors;
+            IntStream.range(0, ackCallbacks.size()).forEach(i -> {
+                final DatastreamRecordMetadata currentRecordMetadata = recordMetadata.get(i);
+                final SendCallback ackCallback = ackCallbacks.get(i);
+                final String topic = currentRecordMetadata.getTopic();
+
+                if (!finalInsertErrors.containsKey(i)) {
+                    DynamicMetricsManager.getInstance().createOrUpdateMeter(classSimpleName, topic, "commitCount", 1);
+                    final long currentRecordSourceTimestamp = sourceTimestamps.get(i);
+                    DynamicMetricsManager.getInstance().createOrUpdateHistogram(classSimpleName, topic, "eteLatency",
+                            System.currentTimeMillis() - currentRecordSourceTimestamp);
+                    ackCallback.onCompletion(currentRecordMetadata, null);
+                } else {
+                    final Exception insertError = finalInsertErrors.get(i);
+                    LOG.warn("Failed to insert a row {} {}", i, insertError.getMessage());
+                    DynamicMetricsManager.getInstance().createOrUpdateMeter(classSimpleName, topic, "errorCount", 1);
+                    ackCallback.onCompletion(currentRecordMetadata, insertError);
+                }
+            });
+
             callback.commited();
         };
 
         _executor.execute(committerTask);
     }
 
+    private Map<Integer, Exception> insertRowsAndMapErrors(final InsertAllRequest insertAllRequest) {
```

Comment:
Looks like the indexes returned by `getInsertErrors()` can be `long` values as mentioned in the [Bigquery API](https://googleapis.dev/java/google-cloud-bigquery/latest/com/google/cloud/bigquery/InsertAllResponse.html#getInsertErrors--)

Instead defining the Map as `Integer` change it to `long`. 
Looks like there many places where you are using  `Map<Integer, Exception>`

Reply:
Hey @r2d2li . True, the Bigquery API does return a map of Long to Exception. The Long in this case is the index of the row from the insert all request. The indexes in Java collections are represented as integers, so it seemed like it would make our lives easier if we stored the insert exceptions as a mapping of Integer to Exception instead, since the index value safely fits within an integer. I wonder if the Longs being returned is a bit of legacy from the library, since the API documentation now describes returning an int, https://cloud.google.com/bigquery/docs/reference/rest/v2/tabledata/insertAll#response-body